File: ios/Classes/AutoOrientationPlugin.m
#import "AutoOrientationPlugin.h"

@implementation AutoOrientationPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"auto_orientation"
                                     binaryMessenger:[registrar messenger]];
    AutoOrientationPlugin* instance = [[AutoOrientationPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    [self setOrientaion:call.method];
    
    [UIViewController attemptRotationToDeviceOrientation];
    
    result(FlutterMethodNotImplemented);
}

- (void)setOrientaion:(NSString *)orientation {
    if (@available(iOS 16.0, *)) {
        NSArray *scenes = [[[UIApplication sharedApplication] connectedScenes] allObjects];
        UIWindowScene *scene = [scenes firstObject];
        
        UIInterfaceOrientationMask orientaionMask = UIInterfaceOrientationMaskAll;
        
        if ([@"setLandscapeRight" isEqualToString:orientation]) {
            orientaionMask = UIInterfaceOrientationMaskLandscapeRight;
        }
        
        if ([@"setLandscapeLeft" isEqualToString:orientation]) {
            orientaionMask = UIInterfaceOrientationMaskLandscapeLeft;
        }
        
        if ([@"setPortraitUp" isEqualToString:orientation]) {
            orientaionMask = UIInterfaceOrientationMaskPortrait;
        }
        
        if ([@"setPortraitDown" isEqualToString:orientation]) {
            orientaionMask = UIInterfaceOrientationMaskPortraitUpsideDown;
        }
        
        if ([@"setPortraitAuto" isEqualToString:orientation]) {
            orientaionMask = UIInterfaceOrientationMaskPortrait;
        }
        
        if ([@"setLandscapeAuto" isEqualToString:orientation]) {
            orientaionMask = UIInterfaceOrientationMaskLandscape;
        }
        
        if ([@"setAuto" isEqualToString:orientation]) {
            orientaionMask = UIInterfaceOrientationMaskAll;
        }
        
        
        UIWindowSceneGeometryPreferencesIOS *geometryPerferencesIOS = [[UIWindowSceneGeometryPreferencesIOS alloc] initWithInterfaceOrientations: orientaionMask];
        [scene requestGeometryUpdateWithPreferences:geometryPerferencesIOS errorHandler:^(NSError * _Nonnull error) {
            NSLog(@"kingiol error: %@", error);
        }];
        
    } else {
        if ([@"setLandscapeRight" isEqualToString:orientation]) {
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationLandscapeRight) forKey:@"orientation"];
        }
        
        if ([@"setLandscapeLeft" isEqualToString:orientation]) {
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationLandscapeLeft) forKey:@"orientation"];
        }
        
        if ([@"setPortraitUp" isEqualToString:orientation]) {
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationPortrait) forKey:@"orientation"];
        }
        
        if ([@"setPortraitDown" isEqualToString:orientation]) {
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationPortraitUpsideDown) forKey:@"orientation"];
        }
        
        if ([@"setPortraitAuto" isEqualToString:orientation]) {
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationPortrait) forKey:@"orientation"];
        }
        
        if ([@"setLandscapeAuto" isEqualToString:orientation]) {
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationLandscapeRight) forKey:@"orientation"];
        }
        
        if ([@"setAuto" isEqualToString:orientation]) {
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationUnknown) forKey:@"orientation"];
        }
    }
}

@end

File: ios/Classes/AutoOrientationPlugin.swift
import Flutter
import UIKit

public class AutoOrientationPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "auto_orientation", binaryMessenger: registrar.messenger())
        let instance = AutoOrientationPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if #available(iOS 16.0, *) {
            setOrientation(call)
        } else {
            setLegacyOrientation(call)
        }

        UIViewController.attemptRotationToDeviceOrientation()

        result(FlutterMethodNotImplemented)
    }
    
    @available(iOS 16.0, *)
    func setOrientation(_ call: FlutterMethodCall) {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene
            else { return }
        let resolvedMask: UIInterfaceOrientationMask
        switch call.method {
        case "setLandscapeRight", "setLandscapeAuto":
            resolvedMask = UIInterfaceOrientationMask.landscapeRight
        case "setLandscapeLeft":
            resolvedMask = UIInterfaceOrientationMask.landscapeLeft
        case "setPortraitUp", "setPortraitAuto":
            resolvedMask = UIInterfaceOrientationMask.portrait
        case "setPortraitDown":
            resolvedMask = UIInterfaceOrientationMask.portraitUpsideDown
        default:
            resolvedMask = UIInterfaceOrientationMask.all
            break
        }
        windowScene.requestGeometryUpdate(.iOS(interfaceOrientations: resolvedMask)) { error in
            // TODO: Perhaps call back to Flutter with an error
        }
    }
    
    func setLegacyOrientation(_ call: FlutterMethodCall) {
        let resolvedOrientation: UIInterfaceOrientation
        switch call.method {
        case "setLandscapeRight", "setLandscapeAuto":
            resolvedOrientation = UIInterfaceOrientation.landscapeRight
        case "setLandscapeLeft":
            resolvedOrientation = UIInterfaceOrientation.landscapeLeft
        case "setPortraitUp", "setPortraitAuto":
            resolvedOrientation = UIInterfaceOrientation.portrait
        case "setPortraitDown":
            resolvedOrientation = UIInterfaceOrientation.portraitUpsideDown
        default:
            resolvedOrientation = UIInterfaceOrientation.unknown
            break
        }
        UIDevice.current.setValue(resolvedOrientation.rawValue, forKey: "orientation")
    }
}
